src/shims/windows/fs.rs

```diff
@@ -463,29 +467,31 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             let io_status_info = io_status_info.clone();
             callback!(
                 @capture<'tcx> {
-                    count: usize,
+                    count: u32,
                     io_status: MPlaceTy<'tcx>,
                     io_status_info: MPlaceTy<'tcx>,
                 }
                 |this, result: Result<usize, IoError>| {
                     match result {
                         Ok(read_size) => {
-                            assert!(read_size <= count);
+                            assert!(read_size <= count.try_into().unwrap());
                             // This must fit since `count` fits.
                             this.write_int(u64::try_from(read_size).unwrap(), &io_status_info)?;
                             this.write_int(0, &io_status)
                         }
                         Err(e) => {
                             this.write_int(0, &io_status_info)?;
-                            this.write_int(error_kind_to_ntstatus(e), &io_status)
+                            this.write_int(e.to_ntstatus(), &io_status)
                         }
                 }}
             )
         };
 
-        desc.write(this.machine.communicate(), buf, count, this, finish)?;
+        desc.write(this.machine.communicate(), buf, count.try_into().unwrap(), this, finish)?;
 
-        // Return whether this was a success. 0 is success.
+        // Return whether this was a success. 0 is a success.
+        // We wrote the return status into the IO_STATUS_BLOCK in the above callback,
+        // just pull it from there.
```

Comment:
This doesn't work -- the entire reason that that's a callback is that `desc.write` may block the current thread, return immediately, and run the callback later!

This function needs to take a `dest` argument to capture the return place, pass that into the callback, and then write the return value that way.

Reply:
This took me a moment to get - for future reference when I'm fixing this: if the callback doesn't happen immediately, we will return the (possibly uninit) value that is in the IO_STATUS_BLOCK we were passed, then before the user code resumes, the callback will happen, write the value into IO_STATUS_BLOCK, then resume execution. We have to pass the return place into the callback to avoid that problem.

This is fairly obvious in retrospect, but it wasn't clear to me until I reread your comment a couple times.

Reply:
Yeah, "blocking" calls in Miri are somewhat funny in how they behave, though given that it is a single-threaded interpreter simulating concurrency, that's the "obvious" architecture.